### PR TITLE
zmq: add option to block/drop samples when high watermark reached

### LINKED ISSUE #6192: ZeroMQ High Water Mark Sample Dropping
### What happened?

When using the ZMQ PUB Sink and ZMQ SUB Source blocks, there is a parameter "High Watermark" that defaults to -1. In ZeroMQ, this sets the corresponding ZMQ_HWM socket option to a value of "0" which means infinite-sized buffers.

http://api.zeromq.org/2-1:zmq-setsockopt

An infinite-sized buffer is not always desired, so a user may specify a "High Watermark" value that in some way limits how much data (not sure how to convert this number into bytes or samples) can be stored in that buffer at any time. This seems like a good option - if a hardware source is producing data and sending it over ZMQ, maybe we want the ability to still back-pressure if the consuming flowgraph cannot keep up.

The actual default behavior of the ZMQ PUB socket is to drop samples (silently!) when this high watermark is reached. I would argue that if a GNURadio user is attempting to set a high-water mark, they want this back pressure behavior and should have an option to select that behavior through an input parameter. This can easily be done by setting the "ZMQ_XPUB_NODROP" socket option.

**tldr;**

**ZeroMQ PUB will drop samples without telling you if you use the high watermark feature.**

Example flowgraph sends a complex sinusoid over ZMQ PUB/ZUB with a high watermark of 1. The SUB side is intentionally throttled so it cannot keep up, samples get dropped.

image

image

### System Information

OS: Ubuntu 20.04
GR Installation Method: Source

### GNU Radio Version

3.11-git (main)

### Specific Version

v3.11.0.0git-231-g463c3477

### Steps to Reproduce the Problem

Set a high watermark on PUB and SUB blocks, then send over more data than can be consumed on the SUB side. See example flowgraph above for a reference.

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by samwhiting**
Signed-off-by: Sam <sam@whitings.org>

## Description
When using the high watermark option in the zeroMQ PUB/SUB blocks, another relevant parameter is left default which silently drops samples when the high watermark is reached. This PR adds a parameter to allow a user to change this behavior to "block" which would result in backpressure. 

More can be read about the default ZMQ behavior at http://api.zeromq.org/4-2:zmq-setsockopt looking for "ZMQ_XPUB_NODROP".

## Related Issue

Fixes #6192 

## Which blocks/areas does this affect?
ZeroMQ PUB and SUB streaming blocks now have additional option to allow using the high watermark in drop/block modes.

## Testing Done
I have not been able to write a deterministic QA test that proves this fix works. The fix has only really been validated by running examples flowgraphs as shown in the related issue. The changes are consistent with ZeroMQ documentation about usage of the high watermark socket options.

## Checklist
- [x] I have read the [CONTRIBUTING document](https://github.com/gnuradio/gnuradio/blob/main/CONTRIBUTING.md).
- [x] I have squashed my commits to have one significant change per commit. 
- [x] I [have signed my commits before making this PR](https://github.com/gnuradio/gnuradio/blob/main/CONTRIBUTING.md#dco-signed)
- [x] My code follows the code style of this project. See [GREP1.md](https://github.com/gnuradio/greps/blob/main/grep-0001-coding-guidelines.md).
- [x] I have updated [the documentation](https://wiki.gnuradio.org/index.php/Main_Page#Documentation) where necessary.
- [ ] I have added tests to cover my changes, and all previous tests pass.

### CODE REVIEW

**Comment by willcode** [comment]
True that in some cases backpressure on the sending side would be useful. This gets complicated, though, because the receiver does not have any way to know whether samples are being dropped, blocked, or neither. The general assumption has been that if the network/stack can't keep up (the usual reason for the sender blocking) then the zmq link isn't going to be useful.

Short story - this could be a good idea, but I think it needs some discussion.

Another idea would be the addition of an overrun tag or other metric to let the client know what's going on.

**Comment by samwhiting** [comment]
Yeah, I'd love to hear what others think or if they've run into problems with this before. It was a problem for me when tagged samples were being dropped, so frequency/timing information was lost.

Also, I did leave the defaults to what they were before so someone instantiating one of these blocks would not need to concern themselves with the HWM or the drop policy if the use case was fine without.

**Comment by willcode** [comment]
@samwhiting Some code formatting stuff (CI won't run tests unless formatting passes). Either run clang-format or taka a look at https://github.com/gnuradio/gnuradio/actions/runs/3139436708/jobs/5100430538

**Comment by jacobagilbert** [comment]
> The general assumption has been that if the network/stack can't keep up (the usual reason for the sender blocking) then the zmq link isn't going to be useful.

I think this is a valid assumption in many cases, but not all. I think this is a useful feature, especially given it doesn't change any behavior by default and is exposed only in the advanced tab. The real concern to me is the lack of deterministic QA, but this is a core ZMQ feature which should come with some confidence.

**Comment by willcode** [comment]
Users might find "On Full Buffer: Drop/Block" or "Non Blocking/Blocking" more understandable for the GRC param. HWM (high watermark) doesn't scream "buffer full".

**Comment by samwhiting** [comment]
Changed to Drop/Block and "On Full Buffer" - good idea.

**Review by marcusmueller — APPROVED** [review]
Nice! Fact is that it only adds defaulting arguments, so should preserve API and abi

**Comment by samwhiting** [comment]
Just added some conditional statements using `USE_NEW_CPPZMQ_SET_GET` for the newer ZMQ API for setting socket options, should hopefully fix the Ubuntu 22 build warnings.

File: gr-zeromq/lib/pub_sink_impl.cc [hunk 1]
```diff
@@ -37,13 +38,16 @@ pub_sink_impl::pub_sink_impl(size_t itemsize,
...
+                             bool drop_on_hwm)
     : gr::sync_block("pub_sink",
                      gr::io_signature::make(1, 1, itemsize * vlen),
                      gr::io_signature::make(0, 0, 0)),
       base_sink_impl(ZMQ_PUB, itemsize, vlen, address, timeout, pass_tags, hwm, key)
 {
-    /* All is delegated */
+    /* Socket option to prevent dropping of samples (backpressure) */
+    int no_drop = (drop_on_hwm == true) ? 0 : 1;
+    d_socket.setsockopt(ZMQ_XPUB_NODROP, &no_drop, sizeof(no_drop));
```
[the remark this answers is not in the document]
Comment (samwhiting) [reply gr-zeromq/lib/pub_sink_impl.cc:53]: Agreed, not the nicest. Just added support for the newer ZMQ API as well which is slightly better.

**Review by jacobagilbert — APPROVED** [review]

**Comment by marcusmueller** [comment]
by the way, my approval was a code-wise thing, I did not want to preempt utility discussion.

**Review by jsallay — APPROVED** [review]
LGTM

**Pull request merged**